### DISCUSSION TITLE
How to get integer index of datetime[64ns]？

### DISCUSSION THREAD

**Discussion by forestbat** [pydata/xarray#7190] [Q&A] [answered]
I have a DataArray which is called `slice_array` like this：
```
<xarray.DataArray 'TM' (TM: 41146)>
array(['2011-08-02T12:00:00.000000000', '2011-08-02T13:00:00.000000000',
       '2011-08-02T14:00:00.000000000', ..., '2022-09-02T19:00:00.000000000',
       '2022-09-02T20:00:00.000000000', '2022-09-02T21:00:00.000000000'],
      dtype='datetime64[ns]')
Coordinates:
  * TM       (TM) datetime64[ns] 2011-08-02T12:00:00 ... 2022-09-02T21:00:00
```
Now I want to find int index of a special time `special_time=2020-10-18T00:00:00.000000000` from the array, but when I try to use this sentence:
`time_index = slice_array.loc[special_time, 'TM']`
It crashed and tells me this:
```
key = (numpy.datetime64('2020-10-18T00:00:00.000000000'), 'TM'), ndim = 1

    def expanded_indexer(key, ndim):
        """Given a key for indexing an ndarray, return an equivalent key which is a
        tuple with length equal to the number of dimensions.
    
        The expansion is done by replacing all `Ellipsis` items with the right
        number of full slices and then padding the key with full slices so that it
        reaches the appropriate dimensionality.
        """
        ……
        if len(new_key) > ndim:
>           raise IndexError("too many indices")
E           IndexError: too many indices
```
So what's my problem? How to solve this?

**Comment by forestbat** [comment]
```
slice_nparray = slice_array.values.flatten()
time_index = np.argwhere(slice_nparray == predict_time)[0][0]
```
can solve my problem, but if you have better answer, please tell me.

**Comment by keewis** [comment, marked as answer]
you'd need to use the correct syntax for [`indexing`](https://docs.xarray.dev/en/stable/user-guide/indexing.html#indexing-with-dimension-names):
```python
time_index = slice_array.loc[{"TM": special_time}]
```
or
```python
time_index = slice_array.sel(TM=special_time)
# or
time_index = slice_array.sel({"TM": special_time})
```
However, that will not give you the *integer index*, it will allow you to select data for a specific time. If you use that to index other arrays I'd suggest to put all of them into a single `Dataset` and call the `sel` / `loc` on that.

If you truly do need the integer index, you can try querying the underlying index object:
```python
slice_array.indexes["TM"].get_loc(special_time)
```
